Source/websocket/JSONWebToken.cpp

```diff
@@ -167,12 +168,33 @@ namespace Web
                 Crypto::SHA256HMAC hash(_key);
 
                 // Extract the signature and convert it to a binary string.
-				uint8_t signature[Crypto::SHA256HMAC::Length];
-                if (Core::URL::Base64Decode(token.substr(pos + 1).c_str(), static_cast<uint16_t>(token.length() - pos - 1), signature, sizeof(signature), nullptr) == sizeof(signature)) {
+		uint8_t signature[Crypto::SHA256HMAC::Length];
+		string signatureSource(token.substr(pos + 1).c_str());
+                TCHAR* signatureOutput = reinterpret_cast<TCHAR*>(ALLOCA(signatureSource.length() * sizeof(TCHAR)));
+
+                if (nullptr == signatureOutput)
+                {
+                    result = false;
+                    return result;
+                }
+                memset (signatureOutput, 0, sizeof (signatureOutput));
+
+                decodedLength = Core::URL::Base64Decode(
+                            token.substr(pos + 1).c_str(),
+                            static_cast<uint16_t>(token.length() - pos - 1),
+                            reinterpret_cast<uint8_t*>(signatureOutput),
+                            static_cast<uint16_t>(signatureSource.length() * sizeof(TCHAR)),
+                            nullptr);
+		    
+                if ( decodedLength == sizeof(signature)) {
```

Comment:
Looking at this line, I think that the essence of this PR is to make sure that the signature decoded from Base64 is not longer than required, but I guess you could have depicted that upfront by the line:  (signatureSource.length() == ((((8 * sizeof(signature))+ 8)/6) ?  Why go through the hassle of converting the full Base64 string if you could have calculated the length of the Base64 representation of a SHA256 sum up front ? Thunder always strives to be Lean, mean and resource constrained ;-) I think that this has an impact of 1 additional if, the comparison in the if is a compiletime calculatable number, so neglectable performance impact...

Reply:
Hi,

Greetings ! Thank you very much for your code review comments !

Changed code as below, It's working fine.

	bool JSONWebToken::ValidSignature(const mode type, const string& token) const 
	{
        bool result = false;

        // Check if the Hash is correct..
        size_t pos = token.find_last_of('.');

        if (pos != string::npos) {
		
            if (type == JSONWebToken::mode::SHA256) {
					
                // Now calculate what we think it should be..
                Crypto::SHA256HMAC hash(_key);

                // Extract the signature and convert it to a binary string.
				uint8_t signature[Crypto::SHA256HMAC::Length];
								
				if ((token.length() - pos) != ((((8 * sizeof(signature))+ 8)/6)))
				{
					return false;
				}
				
                if (Core::URL::Base64Decode(token.substr(pos + 1).c_str(), static_cast<uint16_t>(token.length() - pos - 1), signature, sizeof(signature), nullptr) == sizeof(signature)) {

					hash.Input(reinterpret_cast<const uint8_t*>(token.substr(0, pos).c_str()), static_cast<uint16_t>(pos * sizeof(TCHAR)));
					result = (::memcmp(hash.Result(), signature, sizeof(signature)) == 0);
				}
            }
		}

		return (result);
    }


Newly added code is 

				if ((token.length() - pos) != ((((8 * sizeof(signature))+ 8)/6)))
				{
					return false;
				}
				
Here, Value of ((((8 * sizeof(signature))+ 8)/6)) is 44.

Consider this token "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.aHR0cDovL2xvY2FsaG9zdA.doZcrWnknbAEqIAAeyZ9D8_GJ1cT-XoOJkl-Clm39Rk",

length of "doZcrWnknbAEqIAAeyZ9D8_GJ1cT-XoOJkl-Clm39Rk" is 43 only.

Can i do -1 from ((((8 * sizeof(signature))+ 8)/6)) or Can i add . with "doZcrWnknbAEqIAAeyZ9D8_GJ1cT-XoOJkl-Clm39Rk" length.

Please suggest me.

Reply:
Sorry there is a rounding error in the example, it should be ((((8 * sizeof(signature))+ 7)/6)). This results in ((8 * 32) + 7) / 6 = 43. The +7 is to componsate for the fact that if we need 1 additional byte, it would result in (7+1) a full byte as the divsion will just drop off any remainder, but we need to align on full bytes..

Reply:
Hi,

Greetings ! Thank you very much for your review comments !

Addressed review comments, Added 

                if ((token.length() - pos - 1) == ((((8 * sizeof(signature)) + 7)/6)))
                {

                }
and removed all other newly added lines, Done unit testing, It's working fine.

Please review & merge it.       

Thank you very much once again !